### PR TITLE
Handle dynamic ns elements and tags

### LINKED ISSUE #10: Heka published should handle dynamic metrics in a proper/standard way
The current plugin implementation uses [this Go code](https://github.com/intelsdi-x/snap-plugin-publisher-heka/blob/master/snapheka/snapheka_client.go#L90-L114) for converting a Snap metric into an Heka message.

Here is an example of Heka message produced by this code:

```C
:Uuid: 64727217-9546-4941-BCEB-75EDC2076
:Timestamp: 2016-06-15 06:55:04.415659520 +0000 UTC
:Type: snap.heka
:Logger: snap.heka.logger
:Severity: 6
:Payload:[
   {"namespace":[{"Value":"intel","Description":"","Name":""}
   {"Value":"procfs","Description":"","Name":""},
   {"Value":"disk","Description":"","Name":""},{"Value":"vda1","Description":"","Name":""},
   {"Value":"merged_read","Description":"","Name":""}],
   "last_advertised_time":"0001-01-01T00:00:00Z","version":0,"config":null,"data":0,
   "tags":{"plugin_running_on":"lma-3n24k"},"Unit_":"","description":"","timestamp":"2016-06-15T06:55:04.358782575Z"}]
:EnvVersion: <nil>
:Pid: 135
:Hostname: lma-3n24k
:Fields:
	| name:namespace type:0 representation: value:intel.procfs.disk.vda1.merged_read
	| name:data type:3 representation: value:0
	| name:source type:0 representation: value:unknown_host
	| name:version type:2 representation: value:0
	| name:timestamp type:2 representation: value:1.4659737043588e+18
```

The major problem with this message format is that it doesn’t includes fields for the dynamic elements of dynamic metrics. In our example “vda1” is the value of a dynamic element whose name is “disk”. In InfluxDB we will want to store “disk=vda1” as a tag. For example:

```C
intel.procfs.disk.merged_read,disk=vda1 value=0 1.4659737043588e+18
```

The current Heka message format used does not allow that, so it should be changed.

# Proposed change

**Dynamic Namespace Elements**

Using our previous example, for “disk=vda1” to be used as an InfluxDB tag the Heka message format must be changed. This is what the Heka message should look like:

```C
:Uuid: 64727217-9546-4941-BCEB-75EDC2076
:Timestamp: 2016-06-15 06:55:04.415659520 +0000 UTC
:Type: snap.heka
:Logger: snap.heka.logger
:Severity: 6
:Payload:[
   {"namespace":[{"Value":"intel","Description":"","Name":""},
   {"Value":"procfs","Description":"","Name":""},{"Value":"disk","Description":"","Name":""},
   {"Value":"vda1","Description":"","Name":""},
   {"Value":"merged_read","Description":"","Name":""}],
   "last_advertised_time":"0001-01-01T00:00:00Z","version":0,"config":null,"data":0,
   "tags":{"plugin_running_on":"lma-3n24k"},"Unit_":"","description":"","timestamp":"2016-06-15T06:55:04.358782575Z"}]
:EnvVersion: <nil>
:Pid: 135
:Hostname: lma-3n24k
:Fields:
	| name:name type:0 representation: value:intel.procfs.disk.merged_read
	| name:value type:3 representation: value:0
	| name:timestamp type:2 representation: value:1.4659737043588e+18
	| name:disk type:0 representation: value:vda1
	| name:dimensions type:3 representation: value:disk
```

One the Field is “disk”, whose value is “vda1”. With this we will able to extract InfluxDB tags out of Snap dynamic elements.

**Implementation note**: 

The “dimensions” field will actually be an array of strings. This is how this field can be created and populated using the Heka message API:

```C
field, err := message.NewField(“dimensions”, “foo”, “”)
if err == nil { 
    field.AddValue(“bar”)
    field.AddValue(“baz”)
}
```

This creates a field named “dimensions” whose value is an array of three strings (“foo”, “bar” and “baz”).

**Tags**

The user may set tags in his workflow definition (user-defined tags). For [example using this JSON file](https://github.com/intelsdi-x/snap/blob/7e41c44c284bc5ee411ac01510733016d0d8b148/scheduler/wmap/sample/1.json#L15-L22)

Tags will be handled as dynamic elements, and added as dimension fields to the Heka message. For example:

```C
:Uuid: 64727217-9546-4941-BCEB-75EDC2076
:Timestamp: 2016-06-15 06:55:04.415659520 +0000 UTC
:Type: snap.heka
:Logger: snap.heka.logger
:Severity: 6
:Payload:[
   {"namespace":[{"Value":"intel","Description":"","Name":""},
   {"Value":"procfs","Description":"","Name":""},{"Value":"disk","Description":"","Name":""},
   {"Value":"vda1","Description":"","Name":""},
   {"Value":"merged_read","Description":"","Name":""}],
   "last_advertised_time":"0001-01-01T00:00:00Z","version":0,"config":null,"data":0,
   "tags":{"plugin_running_on":"lma-3n24k"},"Unit_":"","description":"","timestamp":"2016-06-15T06:55:04.358782575Z"}]
:EnvVersion: <nil>
:Pid: 135
:Hostname: lma-3n24k
:Fields:
	| name:name type:0 representation: value:intel.procfs.disk.merged_read
	| name:value type:3 representation: value:0
	| name:timestamp type:2 representation: value:1.4659737043588e+18
	| name:disk type:0 representation: value:vda1
	| name:tag1 type:0 representation: value:val1
	| name:dimensions type:3 representation: value:disk,tag1
```

### PR DESCRIPTION
**Pull request by elemoine**
This PR implements the changes proposed in #10. The dynamic namespace elements and  tags of a metric are now converted into fields in the corresponding Heka message. Unit tests have been added.

Please review.

Fixes #10

### CODE REVIEW

File: snapheka/snapheka_client.go [hunk 1]
```diff
@@ -99,18 +105,71 @@ func snapToHekaPayload(pl string, m plugin.MetricType, pid int32, hostname strin
...
 
-	addField("namespace", strings.Join(m.Namespace().Strings(), "."), msg)
-	addField("data", getData(m.Data()), msg)
-	addField("source", hostname, msg)
-	addField("version", m.Version(), msg)
-	addField("timestamp", m.Timestamp().UnixNano(), msg)
+// Function used to add a specific dynamic metric namespace element
+// into the dimensions field of final Heka message structure
+func addToDimensions(f *message.Field, fName string) (*message.Field, error) {
+	// If the dimension field does not exists yet, create ti
```
Comment (candysmurf) [inline snapheka/snapheka_client.go]: /s/ti/it

**Comment by candysmurf** [comment]
@elemoine, thanks for your work. I'm pulling your PR and testing it now and will update you later.

**Comment by candysmurf** [comment]
@elemoine, were you able to run this plugin against Snap & Heka? I have a trouble to get Snap running after snap-plugin-publisher-heka is built. The error I got was:

_`2016/06/25 21:29:43 proto: duplicate proto type registered: rpc.CollectMetricsReply`_

**Comment by candysmurf** [comment]
@elemoine: The latest Snap is broken. It caused by this merge: e68b1a3df159125502e44efbc007425a7138ea12. I'll use earlier Snap to test your change and update you later.

File: snapheka/snapheka_client.go [hunk 2]
```diff
@@ -99,18 +105,71 @@ func snapToHekaPayload(pl string, m plugin.MetricType, pid int32, hostname strin
...
+				logger.WithField("_block", "fillMetricNameTagsDimensions").Error(err)
+				return err
+			}
+			addField(tag, value, msg)
+		}
+	}
+	if dimField != nil {
+		msg.AddField(dimField)
+	}
+	addField("name", strings.Join(mName, "."), msg)
```
Comment (candysmurf) [inline snapheka/snapheka_client.go]: Do you mind to change "name" to "key" to highlight it's the snap namespace. If you have other idea, please suggest. The reason is that there are "name" field in other collectors, such as Cassandra collector. Interestingly, Heka messages didn't overwrite each other.

**Comment by candysmurf** [comment]
@elemoine, it's fully tested, It worked great! Please see one sample testing output:
```
:Timestamp: 2016-06-27 02:37:59.170732271 +0000 UTC
:Type: snap.heka
:Hostname: egu-mac01.lan
:Pid: 19290
:Uuid: e3469703-276e-4255-b7ac-6a4fe55fba89
:Logger: snap.heka.logger
:Payload: [{"namespace":[{"Value":"intel","Description":"","Name":""},{"Value":"cassandra","Description":"","Name":""},{"Value":"node","Description":"","Name":""},{"Value":"192.168.99.100","Description":"","Name":""},{"Value":"org.apache.cassandra.metrics","Description":"","Name":""},{"Value":"type","Description":"","Name":""},{"Value":"Table","Description":"","Name":""},{"Value":"keyspace","Description":"","Name":""},{"Value":"system","Description":"","Name":""},{"Value":"scope","Description":"","Name":""},{"Value":"size_estimates","Description":"","Name":""},{"Value":"name","Description":"","Name":""},{"Value":"TotalDiskSpaceUsed","Description":"","Name":""},{"Value":"Count","Description":"","Name":""}],"last_advertised_time":"0001-01-01T00:00:00Z","version":0,"config":null,"data":17005,"tags":{"keyspace":"system_auth","name":"RangeLatency","node":"192.168.99.100","plugin_running_on":"egu-mac01.lan","scope":"role_permissions","type":"Table"},"Unit_":"float64","description":"","timestamp":"2016-06-26T19:37:58.187570225-07:00"}]
:EnvVersion:
:Severity: 6
:Fields:
    | name:"keyspace" type:string value:"system_auth"
    | name:"scope" type:string value:"role_permissions"
    | name:"name" type:string value:"RangeLatency"
    | name:"plugin_running_on" type:string value:"egu-mac01.lan"
    | name:"node" type:string value:"192.168.99.100"
    | name:"type" type:string value:"Table"
    | name:"dimensions" type:string value:["keyspace","scope","name","plugin_running_on","node","type"]
    | name:"name" type:string value:"intel.cassandra.node.192.168.99.100.org.apache.cassandra.metrics.type.Table.keyspace.system.scope.size_estimates.name.TotalDiskSpaceUsed.Count"
    | name:"value" type:double value:17005
    | name:"timestamp" type:integer value:1466995078187570225
```

If you can make the commented changes, we can get it merged. Thank you.

File: snapheka/snapheka_client.go (hunk 2 above)
Comment (elemoine) [inline snapheka/snapheka_client.go]: > Do you mind to change "name" to "key" to highlight it's the snap namespace. If you have other idea, please suggest. The reason is that there are "name" field in other collectors, such as Cassandra collector. Interestingly, Heka messages didn't overwrite each other.

Yes, you can have multiple Fields with the same name. It's a feature. On the Heka side I can read the "name" field at index 0, or index 1, etc. See the `fieldIndex` argument to the `read_message` function in the [Heka documentation](http://hekad.readthedocs.io/en/v0.10.0/sandbox/index.html#heka-specific-functions-that-are-exposed-to-the-lua-sandbox).

So relying on that feature we can actually handle collisions by actually having multiple fields with the same name. And our convention is the following: a) the index of a "dimension" field is always 0, and a "standard" field ("name", "value", "timestamp" or "dimensions") is always the last one (its index is actually 0 if there is no collision).

I hope it makes sense.

**Comment by elemoine** [comment]
> If you can make the commented minor changes, we can get it merged. Thank you.

I replied to your comment above on that matter. We can avoid the field collision issue entirely by using multiple fields with the same name, plus some convention for the placement of "dimension" and "standard" fields.

Note I amended my commit to make a few variable/function name changes. Nothing major.

File: snapheka/snapheka_client.go (hunk 2 above)
Comment (candysmurf) [inline snapheka/snapheka_client.go]: @elemoine, ah, nice! It's good to know. thank you.

**Comment by candysmurf** [comment]
LGTM

**Pull request merged**